Deprecate fake packages for lookup. (#3290)
modelica/ModelicaSpecification: chapters/scoping.tex

\item
  If the identifier denotes a class, that class is temporarily flattened (as if instantiating a component without modifiers of this class, see \cref{modification-environment}) and using the enclosing classes of the denoted class.
  The rest of the name (e.g., \lstinline!B! or \lstinline!B.C!) is looked up among the declared named elements of the temporary flattened class.

>   The lookup will only find the element in the following cases:

Reviewer: Looks like a good suggestion, please add it.

  \begin{itemize}

>   \item If the class is declared as \lstinline!package! or \lstinline!operator! (see below) all elements can be found.

Reviewer: Clarify meaning of `operator` (removing "see below" to avoid too much parenthesized content in one place):
```suggestion
  \item If the class is declared as \lstinline!package! or \lstinline!operator! (but not \lstinline!operator record! or \lstinline!operator function!) all elements can be found.
```

\item
  If the identifier denotes a class, that class is temporarily flattened (as if instantiating a component without modifiers of this class, see \cref{modification-environment}) and using the enclosing classes of the denoted class.
  The rest of the name (e.g., \lstinline!B! or \lstinline!B.C!) is looked up among the declared named elements of the temporary flattened class.

>   The restrictions for the lookup are:
>   \begin{itemize}
>   \item If the class is declared as \lstinline!package! or \lstinline!operator! (see below) all elements can be found.
>   \item If the element is declared as \lstinline!encapsulated!\indexinline{encapsulated} it can also be found in other cases.
>   \item A deprecated case is that if the class satisfies the requirements for a \lstinline!package! (without being declared as such), it is still treated as a \lstinline!package!.
>   \item Otherwise it is an error.
>   \end{itemize}

Reviewer: The current formulation about encapsulated sounds as if the rules for encapsulated elements are given by cases described somewhere else.  Also, it's odd that the restrictions come down to listing what is not restricted.
```suggestion
  The lookup may succeed only in the following cases:
  \begin{itemize}
  \item An element can be found if it is declared as \lstinline!encapsulated!\indexinline{encapsulated}.
  \item If the class is declared as \lstinline!package! or \lstinline!operator! (see below) all elements can be found.
  \item A deprecated case is that if the class satisfies the requirements for a \lstinline!package! (without being declared as such), it is still treated as a \lstinline!package!.
  \end{itemize}
```

Author: This seems like a too major change and there are at least two problems with it:

- Lists are often ordered in terms of relevance, and thus leading with the package-case is preferable. The 'also be found in other cases' can be skipped with re-ordering the alternatives.
- 'may succeed only' seems like an odd construct. 'The lookup will only find the element in the following cases: (with slight rewording)

Reviewer: Resolving problems as suggested:
```suggestion
  The lookup will only find the element in the following cases:
  \begin{itemize}
  \item If the class is declared as \lstinline!package! or \lstinline!operator! (see below) all elements can be found.
  \item An element can be found if it is declared as \lstinline!encapsulated!\indexinline{encapsulated}.
  \item A deprecated case is that if the class satisfies the requirements for a \lstinline!package! (without being declared as such), it is still treated as a \lstinline!package!.
  \end{itemize}
```

Reviewer: Regarding "may succeed", I didn't think we wanted to give the impression that the element necessarily exists.  Now, _the element_ sounds as we take its existence for granted, but maybe that's not a big deal as it is obvious that it may not exist?

\item
  If the identifier denotes a class, that class is temporarily flattened (as if instantiating a component without modifiers of this class, see \cref{modification-environment}) and using the enclosing classes of the denoted class.
  The rest of the name (e.g., \lstinline!B! or \lstinline!B.C!) is looked up among the declared named elements of the temporary flattened class.

>   If the class does not satisfy the requirements for a \lstinline!package!, the lookup is restricted to encapsulated\indexinline{encapsulated} elements only.
>   The special case where the class satisfies the requirements for a \lstinline!package!, but is not declared as \lstinline!package! or \lstinline!operator!, is deprecated.

Reviewer: Two things:
1. It's not ideal that the non-deprecated feature is expressed using a sentence about deprecation.
2. Isn't the concept referred to as `operator record` instead of just `operator`?

Put more constructively:
```suggestion
  Unless the class is a \lstinline!package! or \lstinline!operator record!, the lookup is restricted to encapsulated\indexinline{encapsulated} elements only.
  A deprecated semantics is that the lookup restriction does not apply to any class that satisfies the requirements for a \lstinline!package!.
```

Author: Good, except for the 'operator record' part.

There are three different kinds of specialized classes with `operator` in their name: `operator record`, `operator`, and `operator function`, and it is `operator` that is package-like.

As an example:
```
operator record Complex
   Real re,im;
   encapsulated operator '-'
       function negate ... end negate;
       function subtract ... end subtract;
  end '-';
  encapsulated operator function '+' end '+';
end Complex;
```
I can add the rest and an explanation for this.

Author: Have added it.

Reviewer: Looks OK.  Resolving.

>   If the class does not satisfy the requirements for a \lstinline!package!, the lookup is restricted to encapsulated\indexinline{encapsulated} elements only.
>   The special case where the class satisfies the requirements for a \lstinline!package!, but is not declared as \lstinline!package! or \lstinline!operator!, is deprecated.

Reviewer: Two things:
1. It's not ideal that the non-deprecated feature is expressed using a sentence about deprecation.
2. Isn't the concept referred to as `operator record` instead of just `operator`?

Put more constructively:
```suggestion
  Unless the class is a \lstinline!package! or \lstinline!operator record!, the lookup is restricted to encapsulated\indexinline{encapsulated} elements only.
  A deprecated semantics is that the lookup restriction does not apply to any class that satisfies the requirements for a \lstinline!package!.
```

Author: That has now been updated.

Reviewer: OK, let's move on to the new proposal.  Resolving.

  If the identifier denotes a class, that class is temporarily flattened (as if instantiating a component without modifiers of this class, see \cref{modification-environment}) and using the enclosing classes of the denoted class.
  The rest of the name (e.g., \lstinline!B! or \lstinline!B.C!) is looked up among the declared named elements of the temporary flattened class.

>   If the class does not satisfy the requirements for a package, the lookup is restricted to encapsulated\indexinline{encapsulated} elements only.
>   The special case where the class satisfies the requirement for a package, but is not a package, is deprecated.

Reviewer: Would it improve clarity to use the keyword instead of plain text?  Also repeat the plural form of _requirements_ in the added sentence:
```suggestion
  If the class does not satisfy the requirements for a \lstinline!package!, the lookup is restricted to encapsulated\indexinline{encapsulated} elements only.
  The special case where the class satisfies the requirements for a \lstinline!package!, but is not declared as \lstinline!package!, is deprecated.
```

\end{lstlisting}
\end{example}

\begin{nonnormative}
Note that an \lstinline!operator! class may only contain declarations of functions and thus fulfills the requirements for a package (see \cref{specialized-classes}).
In practice, the non-deprecated rules imply that we can call \lstinline!Complex.'-'.negate! and \lstinline!Complex.'+'! for the example in \cref{example-of-overloading-for-complex-numbers}.
This requires that \lstinline!operator '-'! and \lstinline!operator function '+'! are declared as \lstinline!encapsulated! as in the example.

> \end{nonnormative}

Reviewer: Restoring the double empty line spacing befor the `\subsection` to make it easier to find sectioning when scrolling through the source:
```suggestion
\end{nonnormative}

```

\end{lstlisting}
\end{example}

\begin{nonnormative}

> Note that a \lstinline!operator! class may only contain declarations of functions and thus fulfil the requirements for a package (see \cref{specialized-classes}).

Reviewer: Grammar and spelling:
```suggestion
Note that an \lstinline!operator! class may only contain declarations of functions and thus fulfills the requirements for a package (see \cref{specialized-classes}).
```
